### PR TITLE
Improve wanbd logging

### PR DESCRIPTION
**Pull request by eliebak**
# What does this PR do?
need to add a bit more logging like activation and clean this pr, but still usable for now 

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guidelines](https://github.com/huggingface/nanotron/blob/main/CONTRIBUTING.md#submitting-a-new-issue-or-feature-request)?
- [ ] Did you write any new necessary tests?
- [ ] Did you log the throughput and loss you get to ensure the PR works as expected in actual training?
- [ ] Did you log the memory usage? you can use [this tool](https://huggingface.co/spaces/nanotron/predict_memory) to understand the memory usage breakdown in nanotron.
- [ ] If you modified anything related to checkpoints, did you verify that saving and reloading checkpoints still works correctly?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/nanotron/metrics_logging.py [hunk 1]
```diff
@@ -0,0 +1,515 @@
...
+import torch.nn.functional as F
+
+########################
+# Basic utility functions
+########################
+
+
+def compute_tensor_norm(tensor: torch.Tensor, p: int = 2) -> torch.Tensor:
+    """Compute the Lp norm of a tensor."""
+    flattened = tensor.reshape(-1)
```
Comment (NouamaneTazi) [inline src/nanotron/metrics_logging.py:18]: reshape can be expensive at it may require making the tensor contiguous, does torch.norm require contiguous tensors?

File: src/nanotron/metrics_logging.py [hunk 2]
```diff
@@ -0,0 +1,515 @@
...
+
+
+########################
+# Statistical functions
+########################
+
+
+def compute_entropy(tensor: torch.Tensor) -> torch.Tensor:
+    """Compute Shannon entropy of a tensor."""
+    if tensor.min() < 0 or tensor.sum() != 1.0:
```
Comment (NouamaneTazi) [inline src/nanotron/metrics_logging.py:48]: could require CPU-GPU syncs, need to make comparisons on cuda

**Review by NouamaneTazi — CHANGES_REQUESTED** [review]
lgtm overall but left some optimization questions!

File: src/nanotron/metrics_logging.py (hunk 1 above)
Comment (eliebak) [reply src/nanotron/metrics_logging.py]: yep thanks, used torch.linalg.vector_norm(tensor, ord=p) instead

File: src/nanotron/metrics_logging.py (hunk 2 above)
Comment (eliebak) [reply src/nanotron/metrics_logging.py]: removing this function for now (only useful for attention but we don't log it yet) and checked on other should be good

**Comment by eliebak** [comment]
added one wandb run per tp group, will check the tps on multiple run before merging 
Note for later: will have to make it pp compatible later, we can also see how much time and coms it add to log it in one file

File: src/nanotron/metrics_logging.py [hunk 3]
```diff
@@ -15,8 +15,7 @@
 
 def compute_tensor_norm(tensor: torch.Tensor, p: int = 2) -> torch.Tensor:
     """Compute the Lp norm of a tensor."""
-    flattened = tensor.reshape(-1)
-    return torch.norm(flattened, p=p)
+    return torch.linalg.vector_norm(tensor, ord=p)
```
Comment (NouamaneTazi) [inline src/nanotron/metrics_logging.py:18]: nice

File: src/nanotron/metrics_logging.py [hunk 4]
```diff
@@ -43,36 +42,23 @@ def get_attribute_by_path(obj: Any, path: str, debug: bool = False) -> Optional[
...
     """
     Compute kurtosis which measures "tailedness" of distribution.
     High kurtosis indicates outliers, low kurtosis indicates uniform-like distribution.
     """
-    tensor = tensor.flatten().float()
-    mean = tensor.mean()
-    std = tensor.std()
+    # Use float() directly on tensor operations rather than making a copy
+    mean = tensor.float().mean()
+    std = tensor.float().std()
```
Comment (NouamaneTazi) [inline src/nanotron/metrics_logging.py:52]: you're doing tensor.float() too many separate times,
you can instead do 
`tensor = tensor.float()`
in a line above

File: src/nanotron/metrics_logging.py [hunk 5]
```diff
@@ -188,173 +203,86 @@ def compute_global_hidden_layer_metrics(self, model: torch.nn.Module, debug: boo
...
                 
-                metrics[f"{prefix}/mean"] = all_comp_weights.mean()
-                metrics[f"{prefix}/std"] = all_comp_weights.std()
-                metrics[f"{prefix}/min"] = all_comp_weights.min()
-                metrics[f"{prefix}/max"] = all_comp_weights.max()
-                metrics[f"{prefix}/norm_l1"] = compute_tensor_norm(all_comp_weights, p=1)
-                metrics[f"{prefix}/norm_l2"] = compute_tensor_norm(all_comp_weights, p=2)
-                metrics[f"{prefix}/zero_frac"] = compute_zero_fraction(all_comp_weights)
-                metrics[f"{prefix}/kurtosis"] = compute_kurtosis(all_comp_weights)
+            flat_tensors = [w.reshape(-1) for w in weights]
```
Comment (NouamaneTazi) [inline src/nanotron/metrics_logging.py:239]: do we need reshape here :')

**Review by NouamaneTazi — APPROVED** [review]
didn't check everything in details but Ill trust you @eliebak on this one :D 
left some 2 small comments to fix tho

**Comment by eliebak** [comment]
merging, will need to make it compatible with pp (easy) later and maybe do additional coms to sync only on world_rank 0
